File: src/Python/array/75.颜色分类.py
#
# @lc app=leetcode.cn id=75 lang=python
#
# [75] 颜色分类
#

# @lc code=start
class Solution(object):
    def sortColors(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        p0 = 0
        p2 = len(nums) - 1
        # double pointer
        for i in range(p2 + 1):
            # swap untill it does not equal to 2
            while i <= p2 and nums[i] == 2:
                nums[i], nums[p2] = nums[p2], nums[i]
                p2 -= 1
            if nums[i] == 0:
                nums[i], nums[p0] = nums[p0], nums[i]
                p0 += 1
# @lc code=end

File: src/Java/array/75.颜色分类.java
/*
 * @lc app=leetcode.cn id=75 lang=java
 *
 * [75] 颜色分类
 */

// @lc code=start
class Solution {
    public void sortColors(int[] nums) {
        int p0 = 0, p2 = nums.length - 1;
        // double pointer
        for (int i = 0; i <= p2; i++) {
            // swap untill it does not equal to 2
            while (i <= p2 && nums[i] == 2) swap(nums, i , p2--);
            if (nums[i] == 0) swap(nums, i, p0++);
        }
    }
    public void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}
// @lc code=end

